src/data_source/mod.rs

```diff
@@ -170,6 +177,43 @@ impl<'env> Connection<'env> {
     }
 }
 
+impl<'env> Connection<'env, AutocommitOff> {
+    /// Set autocommit mode off, triggers implicit rollback of any running transaction
+    pub fn enable_autocommit(mut self) -> Return<Connection<'env, AutocommitOn>, Self> {
+        match self.handle.rollback() {
+            Success(_) | Info(_) => {},
+            Error(()) => return Error(self.transit()),
+        };
+
+        match self.handle.set_autocommit(true) {
+            Success(_) => Success(self.transit()),
+            Info(_) => Info(self.transit()),
+            Error(()) => Error(self.transit()),
+        }
+    }
+
+    /// Commit transaction if any, can be safely called and will be no-op if no transaction present or autocommit mode is enabled
```

Comment:
One could think about encoding this in the type system as well, i.e. make it statically impossible to commit a transaction if there is none? But since ODBC accepts it, this would probably go beyond odbc-safe's mandate of preventing API errors statically...

Reply:
It looks quite hard to encode this using type system due to autocommit nature of ODBC transactions. Unfortunately we have no explicit transaction start - it will be started with first statement since autocommit was enabled, so I'm not sure when exactly we need to perform type-level connection state transition. Anyway from ODBC point of view commiting when there is no transaction is not error - it's just no-op

Reply:
I was thinking about whenever something is executed, but I think you are right insofar this does not reliably cover all cases and then again it is not an error w.r.t. ODBC.